File: pkg/types/download_job.go
package types

import (
	"encoding/json"
	"sync/atomic"
)

type DownloadJob struct {
	Url             string        `json:"url"`
	Filename        string        `json:"filename,omitempty"`
	Id              string        `json:"id"`
	Password        string        `json:"password,omitempty"`
	Limit           int           `json:"limit,omitempty"`
	ConfigId        string        `json:"configId,omitempty"`
	Transformer     []string      `json:"transformer,omitempty"`
	BytesDownloaded int           `json:"bytesDownloaded,omitempty"`
	Size            int           `json:"size,omitempty"`
	DeleteOnCancel  bool          `json:"deleteOnCancel,omitempty"`
	Speed           float64       `json:"speed,omitempty"`
	Eta             float64       `json:"eta,omitempty"`
	Status          atomic.Value  `json:"-"`
	CancelCh        chan struct{} `json:"-"`
	PauseCh         chan struct{} `json:"-"`
}

func (j *DownloadJob) MarshalJSON() ([]byte, error) {
	type Alias DownloadJob // prevent recursion
	return json.Marshal(&struct {
		Status string `json:"status"`
		*Alias
	}{
		Status: func() string {
			val := j.Status.Load()
			if val == nil {
				return ""
			}
			return string(val.(DownloadState))
		}(),
		Alias: (*Alias)(j),
	})
}

func NewDownloadJob() *DownloadJob {
	job := DownloadJob{CancelCh: make(chan struct{}), PauseCh: make(chan struct{}), Transformer: []string{}}
	job.Status.Store(IDLE)
	return &job
}

type DownloadState string

const (
	IDLE        DownloadState = "idle"
	PAUSED      DownloadState = "paused"
	CANCELED    DownloadState = "canceled"
	DOWNLOADING DownloadState = "downloading"
	DONE        DownloadState = "done"
	ERROR       DownloadState = "error"
)

File: ui/src/types/DownloadJob.ts
import z from 'zod'

export const DownloadJob = z.object({
  url: z.string(),
  filename: z.string().optional(),
  id: z.string(),
  password: z.string().optional(),
  limit: z.number().nonnegative().optional(),
  configId: z.string().optional(),
  transformer: z.array(z.string()).optional(),
  bytesDownloaded: z.number().nonnegative().optional(),
  size: z.number().nonnegative().optional(),
  deleteOnCancel: z.boolean().optional(),
  status: z.literal(['idle', 'paused', 'canceled', 'downloading', 'done', 'error']).optional(),
  speed: z.float64().optional(),
  eta: z.float64().optional(),
  isParent: z.boolean().optional(),
  urls: z.array(z.string()).optional(),
  parentId: z.string().optional(),
  headers: z.record(z.string(), z.string()).optional()
})

export type DownloadJob = z.infer<typeof DownloadJob>
